File: src/main/java/me/syari/ss/discord/internal/handle/SocketHandler.java
package me.syari.ss.discord.internal.handle;

import me.syari.ss.discord.api.utils.data.DataObject;
import me.syari.ss.discord.internal.JDA;
import org.jetbrains.annotations.NotNull;

public abstract class SocketHandler {
    protected final JDA api;
    protected long responseNumber;
    protected DataObject allContent;

    public SocketHandler(JDA api) {
        this.api = api;
    }

    public final synchronized void handle(long responseTotal, @NotNull DataObject object) {
        this.allContent = object;
        this.responseNumber = responseTotal;
        final Long guildId = handleInternally(object.getObject("d"));
        if (guildId != null) getJDA().getGuildSetupController().cacheEvent(guildId, object);
        this.allContent = null;
    }

    protected JDA getJDA() {
        return api;
    }

    protected abstract Long handleInternally(DataObject content);
}

File: src/main/java/me/syari/ss/discord/internal/handle/SocketHandler.kt
package me.syari.ss.discord.internal.handle

import me.syari.ss.discord.api.data.DataObject

abstract class SocketHandler {
    var responseNumber: Long = 0
    var allContent: DataObject? = null

    @Synchronized
    fun handle(responseTotal: Long, dataObject: DataObject) {
        allContent = dataObject
        responseNumber = responseTotal
        val guildId = handleInternally(dataObject.getObject("d"))
        if (guildId != null) GuildSetupController.cacheEvent(guildId, dataObject)
        allContent = null
    }

    abstract fun handleInternally(content: DataObject): Long?
}